lib/portlayer/attach/handle.go

```diff
@@ -0,0 +1,179 @@
+
+	// make sure a spec exists
+	handle.SetSpec(nil)
+
+	return &SerialHandler{
+		handle: handle,
+	}
+}
+
+func (s *SerialHandler) toggle(connected bool) error {
+	ctx := context.Background()
+
+	// get the virtual device list
+	devices, err := s.handle.Container.Devices(ctx)
+	if err != nil {
+		return err
+	}
+
+	// select the virtual serial ports
+	serials := devices.SelectByBackingInfo((*types.VirtualSerialPortURIBackingInfo)(nil))
+	if len(serials) == 0 {
+		return fmt.Errorf("Unable to bind a device with desired backing")
+	}
+	if len(serials) > 1 {
+		return fmt.Errorf("Multiple matches found with desired backing")
+	}
+	serial := serials[0]
+	log.Debugf("Found a device with desired backing: %#v", serial)
+
+	if serial.GetVirtualDevice().Connectable.Connected == connected {
+		log.Debugf("Already in the desired state (connected: %t)", connected)
+		return nil
+	}
+
+	log.Debugf("Setting the connected state to %t", connected)
+	// set the values
+	serial.GetVirtualDevice().Connectable.Connected = connected
+	if connected && s.handle.ExecConfig.Sessions[s.handle.ExecConfig.ID].Attach {
+		log.Debugf("Setting the start connected state to %t", connected)
+		serial.GetVirtualDevice().Connectable.StartConnected = s.handle.ExecConfig.Sessions[s.handle.ExecConfig.ID].Attach
+	}
+
+	config := &types.VirtualDeviceConfigSpec{
+		Device:    serial,
+		Operation: types.VirtualDeviceConfigSpecOperationEdit,
+	}
+	s.handle.Spec.DeviceChange = append(s.handle.Spec.DeviceChange, config)
+
+	return nil
+}
+
+// Join adds 1 network backed, 2 file backed serial port to the caller and configures them
+func (s *SerialHandler) Join() error {
+	defer trace.End(trace.Begin(""))
+
+	// FIXME: THERE MUST BE ANOTHER WAY
+	// following is from Create@exec.go
+	ips, err := net.LookupIP(exec.ManagementHostName)
+	if err != nil {
+		return err
+	}
+
+	if len(ips) == 0 {
+		return fmt.Errorf("No IP found on %s", exec.ManagementHostName)
+	}
+
+	if len(ips) > 1 {
+		return fmt.Errorf("Multiple IPs found on %s: %#v", exec.ManagementHostName, ips)
+	}
+
+	// Tether serial port - backed by network
+	serial := &types.VirtualSerialPort{
+		VirtualDevice: types.VirtualDevice{
+			Backing: &types.VirtualSerialPortURIBackingInfo{
+				VirtualDeviceURIBackingInfo: types.VirtualDeviceURIBackingInfo{
+					Direction:  string(types.VirtualDeviceURIBackingOptionDirectionClient),
+					ServiceURI: fmt.Sprintf("tcp://%s:%d", ips[0], serialOverLANPort),
+				},
+			},
+			Connectable: &types.VirtualDeviceConnectInfo{
+				Connected:         false,
+				StartConnected:    false,
+				AllowGuestControl: false,
+			},
+		},
+		YieldOnPoll: true,
+	}
+	config := &types.VirtualDeviceConfigSpec{
+		Device:    serial,
+		Operation: types.VirtualDeviceConfigSpecOperationAdd,
+	}
+	s.handle.Spec.DeviceChange = append(s.handle.Spec.DeviceChange, config)
+
+	VMPathName := s.handle.Spec.VMPathName()
+	VMName := s.handle.Spec.Spec().Name
+	for _, suffix := range []string{"debug", "log"} {
```

Comment:
I'm not comfortable having the logging tied to the attach mechanism. We should chat in person about why these are rolled in here.